src/context.rs

```diff
@@ -341,12 +343,115 @@ impl<A: Actor> Context<A> {
+                    Either::Right(res.unwrap())
+                }
+                Either::Right((res, other)) => {
+                    addr_recv = other;
+                    broadcast_recv = broadcast_rx.recv_async();
+                    Either::Left(res.unwrap())
+                }
+            };
+
+            // To avoid broadcast starvation, try receive a broadcast here
+            if let Ok(Some(broadcast)) = broadcast_rx.try_recv() {
+                self.tick(Either::Left(broadcast), actor).await;
+            }
+
+            self.tick(msg, actor).await;
+            fut = unfinished;
+        }
+
+        fut.await
+    }
+
+    /// Handle any incoming messages for the actor while running a given future. This is similar to
+    /// [`Context::join`], but will exit if the actor is stopped, returning the future. Returns
+    /// `Ok` with the result of the future if it was successfully completed, or `Err` with the
+    /// future if the actor was stopped before it could complete.
+    ///
+    /// ## Example
+    ///
+    /// ```rust
+    /// # use std::time::Duration;
+    /// # use xtra::prelude::*;
+    /// # use smol::future;
+    /// # struct MyActor;
+    /// # #[async_trait] impl Actor for MyActor { type Stop = (); async fn stopped(self) {} }
+    ///
+    /// struct Stop;
+    /// struct Selecting;
+    ///
+    /// #[async_trait]
+    /// impl Handler<Stop> for MyActor {
+    ///     type Return = ();
+    ///
+    ///     async fn handle(&mut self, _msg: Stop, ctx: &mut Context<Self>) {
+    ///         ctx.stop();
+    ///     }
+    /// }
+    ///
+    /// #[async_trait]
+    /// impl Handler<Selecting> for MyActor {
+    ///     type Return = bool;
+    ///
+    ///     async fn handle(&mut self, _msg: Selecting, ctx: &mut Context<Self>) -> bool {
+    ///         // Actor is still running, so this will return Ok
+    ///         let ans = ctx.select(self, future::ready(1)).await.unwrap();
+    ///         println!("Answer is: {}", ans);
+    ///
+    ///         let addr = ctx.address().unwrap();
+    ///         let select = ctx.select(self, future::pending::<()>());
+    ///         let send = addr.send(Stop).split_receiver();
+    ///         send.await;
+    ///
+    ///         // Actor is no longer running, so this will return Err, even though the future will
+    ///         // usually never complete.
+    ///         select.await.is_err()
+    ///     }
+    /// }
+    ///
+    /// // Will print "Started!", "Goodbye!", "Decided not to keep running", and then "Finally stopping."
+    /// # #[cfg(feature = "with-smol-1")]
+    /// smol::block_on(async {
+    ///     let addr = MyActor.create(None).spawn(&mut xtra::spawn::Smol::Global);
+    ///     assert!(addr.is_connected());
+    ///     assert_eq!(addr.send(Selecting).await, Ok(true)); // Assert that the select did end early
+    /// })
+    ///
+    /// ```
+    pub async fn select<F, R>(&mut self, actor: &mut A, mut fut: F) -> Result<R, F>
```

Comment:
Perhaps `Either` would be a better return value here? I don't think `Result` is appropriate because the actor stopping before isn't really a failure case.

Additionally, this would then move closer to the `futures::select` API.

Reply:
> Perhaps Either would be a better return value here? I don't think Result is appropriate because the actor stopping before isn't really a failure case.

That's fair, but I thought that stopping the actor whilst polling in another handler would probably be something similar to failure.

> Additionally, this would then move closer to the futures::select API.

Yea, this makes sense. I think we can change it to either then.